sni.go

```diff
@@ -17,21 +17,40 @@ package tcpproxy
 import (
 	"bufio"
 	"bytes"
+	"context"
 	"crypto/tls"
 	"io"
 	"net"
+	"strings"
 )
 
 // AddSNIRoute appends a route to the ipPort listener that says if the
 // incoming TLS SNI server name is sni, the connection is given to
 // dest. If it doesn't match, rule processing continues for any
 // additional routes on ipPort.
 //
+// By default, the proxy will route all ACME tls-sni-01 challenges
+// received on ipPort to all SNI dests. You can disable ACME routing
+// with StopACME.
+//
 // The ipPort is any valid net.Listen TCP address.
 func (p *Proxy) AddSNIRoute(ipPort, sni string, dest Target) {
+	cfg := p.configFor(ipPort)
+	if !cfg.stopACME {
+		if len(cfg.acmeTargets) == 0 {
+			p.addRoute(ipPort, (*acmeMatch)(&cfg.acmeTargets))
+		}
+		cfg.acmeTargets = append(cfg.acmeTargets, dest)
+	}
+
 	p.addRoute(ipPort, sniMatch{sni, dest})
 }
 
+// AddStopACMESearch prevents ACME probing of subsequent SNI routes.
```

Comment:
The test demonstrates this:

p.AddSNIRoute(":443", "foo.com", tcpproxy.To("1.2.3.4:443"))
p.AddSNIRoute(":443", "bar.com", tcpproxy.To("2.3.4.5:443"))
p.AddStopACMESearch(":443")
p.AddSNIRoute(":443", "baz.com", tcpproxy.To("3.4.5.6:443"))

With that code, the stack of routes in the config struct for :443 will be: `acmeMatch` against `1.2.3.4:443` and `2.3.4.5:443`, then `sniMatch` against `foo.com`, then `sniMatch` against `bar.com`, then `sniMatch` against `baz.com`.

From the API perspective, yes, AddStopACMESearch is behaving as if it's inserting a "acme stops here" route in the stack. You could implement it like that, but then you'd either have to hardcode some acme behavior as a special case in `Proxy.serveConn`, or have a separate route for each candidate acme target, which all get executed sequentially.

I want to probe all eligible targets in parallel, because in a well-behaving system only one will respond with a non-error. So, to do that, as SNI routes get added, I keep track of whether or not the caller has "inserted" the StopACMESearch route, and if they haven't, I roll each new SNI target up into acmeTargets, and give that whole list to a single acme route that probes all of them in parallel.

The behavior is not particularly obvious, but more obvious ways resulted in either a linear rather than parallel probe of targets, some icky cyclical references between config and route types, or some hardcoded routing logic that didn't live in a self-contained route. Have I missed a better way?